File: tests/e2e/fixtures/small-js-project/utils/error-handler.js
/**
 * Error handling utilities
 * Provides consistent error handling across the application
 */

const config = require('../config');

/**
 * Error types for categorization
 * @enum {string}
 */
const ErrorTypes = {
  VALIDATION: 'ValidationError',
  PARSING: 'ParsingError',
  NETWORK: 'NetworkError',
  TIMEOUT: 'TimeoutError',
  UNEXPECTED: 'UnexpectedError'
};

/**
 * Handles errors in a consistent way
 * @param {Error} error - The error object
 * @param {string} context - The context where the error occurred
 * @returns {Object} - Formatted error information
 */
function handleError(error, context = 'unknown') {
  // Get current environment settings
  const envConfig = config.getEnvironmentConfig();
  
  // Determine error type
  const errorType = determineErrorType(error);
  
  // Standardize error message
  const errorInfo = {
    type: errorType,
    message: error.message,
    context,
    timestamp: new Date().toISOString()
  };
  
  // Add stack trace in non-production environments
  if (envConfig.logLevel === 'debug') {
    errorInfo.stack = error.stack;
  }
  
  // Log the error based on environment settings
  logError(errorInfo, envConfig);
  
  return errorInfo;
}

/**
 * Determines the type of error
 * @param {Error} error - The error to categorize
 * @returns {string} - The error type
 */
function determineErrorType(error) {
  if (error.name === 'SyntaxError' || error.message.includes('parse')) {
    return ErrorTypes.PARSING;
  }
  
  if (error.message.includes('validate') || error.message.includes('invalid')) {
    return ErrorTypes.VALIDATION;
  }
  
  if (error.code === 'ECONNREFUSED' || error.code === 'ENOTFOUND') {
    return ErrorTypes.NETWORK;
  }
  
  if (error.code === 'ETIMEDOUT' || error.message.includes('timeout')) {
    return ErrorTypes.TIMEOUT;
  }
  
  return ErrorTypes.UNEXPECTED;
}

/**
 * Logs the error based on environment configuration
 * @param {Object} errorInfo - Formatted error information
 * @param {Object} envConfig - Environment configuration
 */
function logError(errorInfo, envConfig) {
  if (envConfig.enableVerboseLogging) {
    console.error('ERROR DETAILS:', JSON.stringify(errorInfo, null, 2));
  } else {
    console.error(`Error [${errorInfo.type}]: ${errorInfo.message} (in ${errorInfo.context})`);
  }
  
  // In a real application, you might log to a file or service here
}

module.exports = {
  handleError,
  ErrorTypes
};

File: tests/e2e/fixtures/python-project/utils/error_handler.py
"""
Error handling module for the Python test project
"""

import logging
from typing import Dict, Any, Optional

logger = logging.getLogger(__name__)

class ProcessingError(Exception):
    """Base class for all processing errors"""
    def __init__(self, message: str, error_type: str = "ProcessingError", context: Optional[Dict[str, Any]] = None):
        super().__init__(message)
        self.error_type = error_type
        self.context = context or {}

class ValidationError(ProcessingError):
    """Error raised when validation fails"""
    def __init__(self, message: str, context: Optional[Dict[str, Any]] = None):
        super().__init__(message, "ValidationError", context)

class ParsingError(ProcessingError):
    """Error raised when parsing fails"""
    def __init__(self, message: str, context: Optional[Dict[str, Any]] = None):
        super().__init__(message, "ParsingError", context)

class TimeoutError(ProcessingError):
    """Error raised when processing times out"""
    def __init__(self, message: str, context: Optional[Dict[str, Any]] = None):
        super().__init__(message, "TimeoutError", context)

class NetworkError(ProcessingError):
    """Error raised for network issues"""
    def __init__(self, message: str, context: Optional[Dict[str, Any]] = None):
        super().__init__(message, "NetworkError", context)

def handle_error(error: Exception, context_name: str) -> Dict[str, Any]:
    """
    Handle an error and format it consistently
    
    Args:
        error: The exception that occurred
        context_name: The context where the error occurred
        
    Returns:
        Formatted error information
    """
    # Extract error information
    if isinstance(error, ProcessingError):
        error_type = error.error_type
        error_context = error.context
    else:
        error_type = error.__class__.__name__
        error_context = {}
    
    # Format the error
    error_info = {
        "type": error_type,
        "message": str(error),
        "context": context_name,
        **error_context
    }
    
    # Log the error
    logger.error(f"Error in {context_name}: {error_type} - {error}")
    
    return error_info
